MIP/mip-draft-FFS/README.md

```diff
@@ -0,0 +1,229 @@
+
+**Slashing**. The security of the mechanism relies on a PoS protocol. Each validator has to stake some assets, and if they are malicious they _should_ be slashed.
+The condition for slashing may be met by several quiteria, and not all slashing conditions may be used:
+- equivocate (send a different vote to different validators or users)
+- vote :white_check_mark: for an invalid block 
+- vote :x: for a valid block
+
+### Main challenges
+
+To achieve crypto-economically secured Fast-Finality, we need to solve the following problems:
+
+1. design a _staking_ mechanism for the validators to stake assets, distribute rewards and manage slashing
+1. _define and verify_ the threshold (e.g. 2/3 of validators attest :white_check_mark:) for L2-finality
+1. _communicate_ the L2-finality status.
+
+In addition the following may require separate discussion, as it is a different procedure (postconfirmations are handled in smart contracts on L1, whereas L2-finality is handled off-chain)
+
+4. _define and verify_ the threshold (e.g. 2/3 of validators attest :white_check_mark:) for postconfirmation.
+
+
+### Components
+
+#### Staking (addresses 1. and 2.)
+
+The staking mechanism is implemented in a contract on L1.
+This contract provides the following functionalities:
+
+- join: a new validator can join the set of validators by staking some assets
+- exit: a validator exits and  get their stakes back
+- vote: receive a vote or a set of votes, verify the integrity of the votes (signatures) and the minimum threshold (e.g. 2/3)
+
+#### Generate an L2-finality certificate (addreses 3.)
+
+To ensure that the L2-finality status is made available to third-parties, we may publish our _proof_  (2/3 of attestations :white_check_mark:) to a data availability layer and get a _certificate_ that the proof is available.  
+This DA layer should offer a reliable _mempool_ for example as described [in this paper](https://arxiv.org/pdf/2105.11827).
+
+#### Handle postconfirmations (addresses 4.)
+
+The L1 contract will verify the L2-finality certificate. If the certificate is correct the block (or sequence of blocks) are _postconfirmed_. This requires handling who should send the certificate to the L1 contract, and how to verify the certificate.
+
+
+## Reference Implementation
+
+To simplify we assume that each validator stakes the same amount.
+The set of validators is in charge of validating a block at each round.
+
+A leader validator $V_l$ is elected in each round, and proposes the next transition (and block $B'$):  $B \xrightarrow{\ txs \ } B'$.
+They can do by sending a digest of $txs$ (Merkle root) and a digest of $B'$ (Merke root hash of $B'$), or a _change set_.
+Every other validator checks the validity of $B'$ and send their vote to $V_l$.
```

Comment:
This seems overly complicated. Since the root is deterministic no proposer is required. Instead nodes should send the (sequence of) unconfirmed block hashes (that contain the state root) to all their peers. This is unproblematic for throughput as nodes only send signed messages containing a hash (N-to-N). Any validator then can create a quorum certificate for a given block height. 

A leader is required to determine who should update the L1 contract to achieve postconfirmation (which is more worth than L2-finality certificate). There is a liveness question if the leader is inactive, but this seems resolved by rotating the leader every postconfirmation.

Reply:
Let me try to clarify a few points:
- the execution is deterministic, that's true, but some validators can be malicious. So if we give all the validators the source block and the Txs to be processed from this block, we can still have up to $f + 1$ (number of malicious) proposals for the next block. And each of the validator is going to send around their proposals.  We don;'t have a voting/quorum anymore but need a consensus. Correct? 
- if a leader is elected, we have a yes/no question to answer (not a consensus). Is the proposed next block correct? So validators attest for one block in each round. As the leader is also responsible for sending the aggregate attestations and the new block root, we can detect whether they are malicious.
- the leader can inactive (not sending anything intentionally or not) and we can indeed rotate them, Worst case we need $f + 1$ epochs to deliver the attestations. 
- a post confirmation is an event (a super-majority is obtained) but we need to make it visible to third-parties.Thisis the purpose of the availability certificate.

Reply:
1) 
> We don;'t have a voting/quorum anymore but need a consensus. Correct?

I think we don't need a consensus on L2 for a certain proposer block. Here is my idea. 
So the hash of the block has no relation to the sender. The wrapping tx has.. So naively on L2  there is no leader requirement to progress on blocks for a validator to prove to a client a new state is correct. Only 2/3 messages need to be shown.

I am not certain about aggregations in this context. Happy to learn this is an issue.

2) 
With regards to availability to third party. The basic DA is the validator network :) but if that's not enough any validator could store a proof of a new correct state by naively storing the 2/3 messages on a DA

Reply:
3)
With regards to the process involving L1 i agree it gets more tricky. A leader is likely necessary 

4) 
I really think we should clearly distinguish between the event of reaching 2/3 majority on L2 and the event of verifying 2/3 on L1. They are different as protocol, in time, and potentially as actors. They even serve different purposes. In our blog I understood postconfirmation are defined as the latter. Hence I would  advocate for having two different names.